pyquil/paulis.py

```diff
@@ -407,6 +414,74 @@ def pauli_string(self, qubits=None):
 
+    def matrix(self, qubit_mapping={}, nqubits=None, check_mapping=True):
+        """Create the matrix representation of self.
+
+        :param nqubits:  Number of qubits to represent self on. If none the maximum qubit in the
+                         PauliTerm or qubit_mapping is used.
+        :param qubit_mapping: A dictionary-like object that maps from :py:class`QubitPlaceholder`
+                              to :py:class:`int`
+        :check_mapping: (bool) whether or not to check qubit_mapping for validity
+        :return: np.matrix representing the PauliTerm
+        """
+        qubits = self.get_qubits()
+
+        if check_mapping:
+            real_qubits = set()
+            placeholder_qubits = set()
+            for q in qubits:
+                if isinstance(q, int):
+                    real_qubits.add(q)
+                elif isinstance(q, QubitPlaceholder):
+                    placeholder_qubits.add(q)
+                else:
+                    raise ValueError(
+                        "the qubit {} is neither int nor QubitPlaceholder".format(q))
+
+            mapped_placeholders = set(qubit_mapping.keys())
+            mapped_qubits = set(qubit_mapping.values())
+
+            assert placeholder_qubits <= mapped_placeholders,\
+                "not all QubitPlaceholders are mapped to real qubits"
+            assert len(mapped_qubits) == len(mapped_placeholders),\
+                "qubit_mapping is not one to one"
+            assert len(mapped_qubits & real_qubits) == 0,\
+                "qubit_mapping maps to qubits already in use"
+
+        reversed_mapping = {}
+        for qubit in qubits:
+            if isinstance(qubit, int):
+                reversed_mapping[qubit] = qubit
+            elif isinstance(qubit, QubitPlaceholder):
+                try:
+                    reversed_mapping[qubit_mapping[qubit]] = qubit
+                except KeyError:
+                    raise ValueError(
+                        "{} not found in qubit_mapping".format(qubit))
+            else:
+                raise ValueError(
+                    "qubit {} is neither int nor QubitPlaceholder".format(qubit))
+
+        def get_op(qubit_index):
+            """Return the operator at qubit `qubit_index`
+            """
+            try:
+                return PAULI_MATS[self[reversed_mapping[qubit_index]]]
+            except KeyError:  # all qubits w/o an operator have "I" implicitly
+                return(PAULI_MATS["I"])
+
+        # check that nqubits makes sense
+        if nqubits is None:
+            nqubits = max(reversed_mapping.keys()) + 1
+        elif nqubits <= max(reversed_mapping.keys()):
+            raise ValueError(
+                "nqubits must be higher than the largest qubit in the PauliTerm or the qubit_map")
+
+        matrix = np.array([self.coefficient])
+        for i in range(nqubits):
+            matrix = np.kron(get_op(i), matrix)
+        return matrix
```

Comment:
This is a little dodgy. Using this logic, if I were to get the matrix rep of something like `sZ(0) * sX(2)`, I would get an 8 x 8 matrix, which is assuming that qubit `1` is a valid qubit and which may not always be the case, depending on the actual chip/lattice we're running on. I would suggest re-factoring the code so that only the qubits being operated on appear in the matrix representation, and such that the lowest integer-index appears on the right-most, as in the convention of the `WavefunctionSimulator`: http://docs.rigetti.com/en/stable/wavefunction_simulator.html#multi-qubit-basis-enumeration, and one which this code appears to be respecting.

Reply:
This function is called by PauliSum().matrix() to construct the matrix of a PauliSum by summing up the matrices of all terms in the sum. But not all terms will neccesarily act on all qubits that the whole sum acts on. So we have to have factors of the identity in the kronecker product!

(Consider H = X0ZX2 + Z0Z1Z2. If I skip the first qubit in the first term, it will result in a 4x4 matrix while the second term gives a 8x8 matrix which can't be summed.)

Another question is, how to think about this problem on the level of the PauliSums (not PauliTerms). I mostly used this .matrix() function in conjunction with the `WavefunctionSimulator` to easily compute expectation values. So I agree that the aim should be here that the matrix comes out in such a way that it plays nicely together with the wavefunction from the `WavefunctionSimulator`.

Reply:
Since you have separate `matrix` methods for `PauliTerms` and `PauliSums`, I would argue that it makes sense to restrict to the qubit subspace appearing in the `PauliTerm` for its matrix representation. To avoid confusion between e.g. `sX(0) * sZ(2)` and `sX(1) * sZ(3)`, one could access `PauliTerm.get_qubits()`.